Add option to re-indent text after exchange

wilywampa opened the pull request:
This change adds an option to re-indent text after the exchange for linewise exchanges. The option is off by default and can be enabled per buffer.

wellle commented:
I often run into this issue when inverting an if condition and exchanging the if body with the rest of the function. Would use :+1:

tommcdo commented:
I've been thinking of implementing this for along time, but I don't want it to use `=`. It often does more than just shifting left or right.

I'd much prefer that we detect the indentation level of each block and manually adjust them after exchanging. Nothing but leading indentation should be affected.

(I have this issue with surround.vim as well: if I surround a linewise block with `{`, it changes my block of code substantially. I'd prefer it just did an `>` operation on the surrounded text. This is probably because I haven't configured `'indentexpr'`.)

wilywampa commented:
That makes sense and I think I addressed your concerns plus fixed some bugs in the first commit with the commit I just pushed.

tommcdo commented:
The new implementation is still using `=`, even if only for a single line. I'd prefer using the indentation level of the text that was there previously instead of the indentation level that `=` thinks it should use.

After that, if you can squash all of the commits together I'll be happy to merge this in :)

wilywampa commented:
I'll add that as a option but I don't think it makes as much sense. Re-indenting the first line using `==` works well because it is indented in the context of its new position and preserves manual alignment of lines after the first line that would be lost by using `==` on all lines. Just switching the indents of each of the blocks based on the indents of the first lines doesn't work so well. For example, moving a line outside of an `if` statement:

```c
if (foo == 1)
{
    Something();
    foobar = 2 + 3
           + 4 + 5;
}
```
with no change in indentation:
```c
if (foo == 1)
{
    Something();
}
    foobar = 2 + 3  // shouldn't be indented
           + 4 + 5;
```
with my initial PR:
```c
if (foo == 1)
{
    Something();
}
foobar = 2 + 3
    + 4 + 5; // + should align with =
```
with my latest PR:
```c
if (foo == 1)
{
    Something();
}
foobar = 2 + 3
       + 4 + 5;
```
without `==`:
```c
if (foo == 1)
{
    Something();
    } // shouldn't be indented
foobar = 2 + 3
       + 4 + 5;
```

tommcdo commented:
I don't consider that a use-case for exchange.vim. What you're doing is moving one piece of text, not exchanging two pieces of text.

Here's a use case where I think `=` is totally wrong:

```c
fooThing = a + b
         + somethingForBar();
if (condition) {
    barThing = c + d
             + somethingForFoo();
}
```

Exchanging the `somethingFor*` lines here, `=` would destroy the alignment.

wilywampa commented:
I implemented swapping the existing indent of each region and squashed/rebased on master. It got a bit more complicated to handle tabs/spaces/mixed and blank lines that didn't work properly before. I put in an option to re-indent using `==` on the first line of each region by setting `g:exchange_indent = '=='`. Hope that's okay. Neither mode is enabled by default.

tommcdo commented:
Sorry for taking so long to get back to you on this.

I think there are some bugs to work out. Using `g:exchange_indent = 1`, a lot of my testing left incorrect indentation. Often, it seemed that one of the regions was indented correctly, while the other had no indentation at all.

Another thing that didn't work was a region with another region fully contained inside of it. For example, replacing an `if` statement with just the body (i.e. remove the conditional part). It seems that some of the math used affects the indentation of nearby lines.

I'm gonna leave this open, and over time we can figure out where the bugs are and squash them.

wilywampa commented:
I'm glad you're testing it instead of blindly merging. Can you be more specific with your examples?

wilywampa commented:
I fixed the overlapping region case. I'll squash the commits again after we work out all the kinks.

tommcdo commented:
I finally got around to testing these new changes. I'm gonna go ahead and merge this in. It seems to be in working order.

Thanks for adding this, and sorry for my slow responses!
